avogadro/qtgui/layermodel.cpp

```diff
@@ -0,0 +1,195 @@
+/******************************************************************************
+  This source file is part of the Avogadro project.
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#include "layermodel.h"
+#include "molecule.h"
+#include "rwmolecule.h"
+
+#include <QtCore/QFileInfo>
+#include <QtGui/QColor>
+#include <QtGui/QIcon>
+
+namespace Avogadro {
+namespace QtGui {
+
+using Core::Layer;
+
+const int LayerModel::QTTY_COLUMNS = 6;
+
+LayerModel::LayerModel(QObject* p) : QAbstractItemModel(p), m_item(0) {}
+
+QModelIndex LayerModel::parent(const QModelIndex&) const
+{
+  return QModelIndex();
+}
+
+int LayerModel::rowCount(const QModelIndex& p) const
+{
+  if (p.isValid())
+    return 0;
+  else
+    return m_item;
+}
+
+int LayerModel::columnCount(const QModelIndex& p) const
+{
+  return QTTY_COLUMNS;
+}
+
+Qt::ItemFlags LayerModel::flags(const QModelIndex& idx) const
+{
+  return Qt::ItemIsEnabled;
+}
+
+bool LayerModel::setData(const QModelIndex& idx, const QVariant& value,
+                         int role)
+{
+  return false;
+}
+
+QVariant LayerModel::data(const QModelIndex& idx, int role) const
+{
+  if (!idx.isValid() || idx.column() > QTTY_COLUMNS)
+    return QVariant();
+  auto names = activeMoleculeNames();
+  if (idx.row() == names.size()) {
+    if (idx.column() == 0) {
+      switch (role) {
+        case Qt::DecorationRole:
+          return QIcon(":/icons/fallback/32x32/plus.png");
+        default:
+          return QVariant();
+      }
+    }
+    return QVariant();
+  }
+  auto name = names[idx.row()].second.toStdString();
+  auto layer = names[idx.row()].first;
+  bool isLayer = name == tr("Layer").toStdString();
+  if (isLayer) {
+    if (idx.column() == 0) {
+      switch (role) {
+        case Qt::DisplayRole: {
+          return (name + " " + std::to_string(layer)).c_str();
+        }
+        case Qt::ForegroundRole:
+          if (layer == static_cast<int>(getMoleculeLayer().activeLayer()))
+            return QVariant(QColor(Qt::red));
+          else
+            return QVariant(QColor(Qt::black));
+        default:
+          return QVariant();
+      }
+    } else if (idx.column() == (QTTY_COLUMNS - 5)) {
```

Comment:
I'd declare an enum  -- either in the header or here. That way you can have `(idx.column()  == LayerVisible)` and it makes it much easier to change / add  columns in the future.

e.g., `propertymodel.*`
https://github.com/OpenChemistry/avogadrolibs/pull/711/files